File: RepoCoderStudio/repo_explorer_data/sample_repo/accounts/KycValidator.java
package accounts;

import java.util.HashMap;
import java.util.Map;

/**
 * Customer identity verification (KYC) for the demo BFSI sample repository.
 *
 * Java counterpart of accounts/kyc_validator.py -- same checks, kept in
 * step with the Python version so Stage 4/5 has a genuinely bilingual
 * repository to index rather than a Python-only demo.
 */
public class KycValidator {

    /**
     * Check whether a string matches the Indian PAN card format (5
     * letters, 4 digits, 1 letter).
     */
    public boolean validatePan(String panNumber) {
        String normalized = panNumber.trim().toUpperCase();
        return normalized.matches("[A-Z]{5}[0-9]{4}[A-Z]{1}");
    }

    /**
     * Check whether a string is a plausible 12-digit Aadhaar number.
     */
    public boolean validateAadhaar(String aadhaarNumber) {
        String digits = aadhaarNumber.replace(" ", "");
        return digits.matches("[0-9]+") && digits.length() == 12;
    }

    /**
     * Run full KYC identity verification, combining PAN and Aadhaar
     * checks.
     */
    public Map<String, Boolean> verifyIdentity(String panNumber, String aadhaarNumber, String fullName) {
        boolean panOk = validatePan(panNumber);
        boolean aadhaarOk = validateAadhaar(aadhaarNumber);
        boolean nameOk = fullName != null && !fullName.trim().isEmpty();

        Map<String, Boolean> result = new HashMap<>();
        result.put("verified", panOk && aadhaarOk && nameOk);
        result.put("panValid", panOk);
        result.put("aadhaarValid", aadhaarOk);
        result.put("nameProvided", nameOk);
        return result;
    }

    /**
     * Mask a PAN number for display, keeping only the last 4 characters
     * visible.
     */
    public String maskPan(String panNumber) {
        String normalized = panNumber.trim().toUpperCase();
        if (normalized.length() < 4) {
            return "*".repeat(normalized.length());
        }
        String visible = normalized.substring(normalized.length() - 4);
        return "*".repeat(normalized.length() - 4) + visible;
    }
}

File: RepoCoderStudio/repo_explorer_data/sample_repo/accounts/kyc_validator.py
"""Customer identity verification (KYC) for the demo BFSI sample repository."""

from __future__ import annotations

import re


def validate_pan(pan_number: str) -> bool:
    """Check whether a string matches the Indian PAN card format (5 letters, 4 digits, 1 letter)."""
    return bool(re.fullmatch(r"[A-Z]{5}[0-9]{4}[A-Z]{1}", pan_number.strip().upper()))


def validate_aadhaar(aadhaar_number: str) -> bool:
    """Check whether a string is a plausible 12-digit Aadhaar number."""
    digits = aadhaar_number.replace(" ", "")
    return digits.isdigit() and len(digits) == 12


def verify_identity(pan_number: str, aadhaar_number: str, full_name: str) -> dict:
    """Run full KYC identity verification, combining PAN and Aadhaar checks."""
    pan_ok = validate_pan(pan_number)
    aadhaar_ok = validate_aadhaar(aadhaar_number)
    name_ok = bool(full_name and full_name.strip())
    return {
        "verified": pan_ok and aadhaar_ok and name_ok,
        "pan_valid": pan_ok,
        "aadhaar_valid": aadhaar_ok,
        "name_provided": name_ok,
    }


def mask_pan(pan_number: str) -> str:
    """Mask a PAN number for display, keeping only the last 4 characters visible."""
    pan_number = pan_number.strip().upper()
    if len(pan_number) < 4:
        return "*" * len(pan_number)
    return "*" * (len(pan_number) - 4) + pan_number[-4:]
